galaxy/models.py

```diff
@@ -118,60 +135,19 @@ class Galaxy(models.Model):
     PICTURE_POINTS = 2  # Equivalent to two days as random members of a club.
     CLUBS_POINTS = 1  # One day together as random members in a club is one point.
 
-    state = models.JSONField("current state")
+    state = models.JSONField(_("The galaxy current state"), null=True)
 
-    @staticmethod
-    def make_state() -> None:
-        """
-        Compute JSON structure to send to 3d-force-graph: https://github.com/vasturiano/3d-force-graph/
-        """
-        without_nickname = Concat(
-            F("owner__first_name"), Value(" "), F("owner__last_name")
-        )
-        with_nickname = Concat(
-            F("owner__first_name"),
-            Value(" "),
-            F("owner__last_name"),
-            Value(" ("),
-            F("owner__nick_name"),
-            Value(")"),
-        )
-        stars = GalaxyStar.objects.annotate(
-            owner_name=Case(
-                When(owner__nick_name=None, then=without_nickname),
-                default=with_nickname,
-            )
-        )
-        lanes = GalaxyLane.objects.annotate(
-            star1_owner=F("star1__owner__id"),
-            star2_owner=F("star2__owner__id"),
-        )
-        json = GalaxyDict(
-            nodes=[
-                StarDict(id=star.owner_id, name=star.owner_name, mass=star.mass)
-                for star in stars
-            ],
-            links=[],
-        )
-        # Make bidirectional links
-        # TODO: see if this impacts performance with a big graph
-        for path in lanes:
-            json["links"].append(
-                {
-                    "source": path.star1_owner,
-                    "target": path.star2_owner,
-                    "value": path.distance,
-                }
-            )
-            json["links"].append(
-                {
-                    "source": path.star2_owner,
-                    "target": path.star1_owner,
-                    "value": path.distance,
-                }
-            )
-        Galaxy.objects.all().delete()
-        Galaxy(state=json).save()
+    class Meta:
+        ordering = ["pk"]
```

Comment:
You could also add a [constraint](https://docs.djangoproject.com/en/3.2/ref/models/constraints/) to make sure there cannot be more than one galaxy with a state (or two, if you want to generate a galaxy, then delete the previous one)

Reply:
I'm not sure that's such a good idea. Constraints are great to enforce certain conditions that we really want to see respected, and thus can often lead to exceptions being thrown out of nowhere in case of a bug going against the constraint. In the case of the Galaxy, a cron job failing in the middle of the ruling, or any other unexpected error could be even worse if we add such a constraint, when it would only have been cleaned up peacefully with a later run of the `rule_galaxy` command, that already takes care of garbage collection.